### PR TITLE
[CodeCompletion] Migrate expression completions to solver-based

### PR DESCRIPTION
**Pull request by ahoppen**
Migrate `StmtOrExpr` completions to solver-based.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -9880,6 +9880,11 @@ bool ConstraintSystem::resolveClosure(TypeVariableType *typeVar,
     }
   }
 
+  SolutionApplicationTarget target(closure, closure,
+                                   ContextualTypePurpose::CTP_CallArgument,
+                                   contextualType, /*isDiscarded=*/false);
+  setSolutionApplicationTarget(closure, target);
```
Comment (ahoppen) [inline lib/Sema/CSSimplify.cpp:9886]: @xedin I’m pretty sure this part is not correct. As we discussed, I’m using `SolutionApplicationTarget` to convey the contextual type of the `ClosureExpr`. I’m not entirely sure what you’d expect for the decl context, contextual type purpose and `isDiscarded`. AFAICT, we don’t really know the contextual type purpose and whether it’s discarded inside `resolveClosure`.

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: include/swift/Sema/ConstraintSystem.h [hunk 2]
```diff
@@ -1386,6 +1386,13 @@ class Solution {
   /// "resolved" concrete type.
   Type getResolvedType(ASTNode node) const;
 
+  /// Retrieve the type of the given node as recorded in this solution
+  /// and resolve all of the type variables in contains to form a fully
+  /// "resolved" concrete type.
+  /// If no type has been computed for this node or if \c node is null, returns
+  /// the null type.
+  Type getResolvedTypeIfAvailable(ASTNode node) const;
```
Comment (xedin) [inline include/swift/Sema/ConstraintSystem.h:1394]: I don't see any uses of this but I'd say that if ASTNode doesn't have a type it's either not associated with the solution which is an issue on the caller side, or it didn't get recorded during constraint generation, which makes it a bug in the solver.

**Review by xedin — COMMENTED** [review]
Looked only at the last commit.

File: lib/Sema/CSSimplify.cpp (hunk 1 above)
Comment (ahoppen) [reply lib/Sema/CSSimplify.cpp:9886]: Thanks. I changed it.

File: include/swift/Sema/ConstraintSystem.h (hunk 2 above)
Comment (ahoppen) [reply include/swift/Sema/ConstraintSystem.h:1394]: Oh, looks like the need for that method has gone away at some point during my development. I just removed it.

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
Stress tester run: https://ci.swift.org/job/swift-PR-macos-with-sourcekit-stress-tester/44/

This needs some XFails due to SR-16012 and a performance problem I’m addressing in separately in https://github.com/apple/swift/pull/41917.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Review by xedin — APPROVED** [review]

**Pull request merged**